### PR TITLE
[AIR/Docs] GPT-J fine tuning with DeepSpeed example

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Adds an example.

Depends on https://github.com/ray-project/ray/pull/33089, https://github.com/ray-project/ray/pull/33088, https://github.com/ray-project/ray/pull/32955

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb [hunk 1]
```diff
@@ -0,0 +1,1151 @@
...
+     "execution_count": 5,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "import ray\n",
+    "\n",
+    "ray.init(\n",
+    "    runtime_env={\n",
```
Comment (ericl) [inline doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb:149]: From a separate discussion, I think we should generally avoid using runtime_env in our examples, as they are a very advanced concept. We can suggest `pip install --user` instead, which will work seamlessly in workspaces.

Looking at the rest of our examples it seems we don't use runtime_env there, so maybe we can remove this from our recent examples.

cc @richardliaw

Comment (Yard1) [reply doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb:149]: One issue is that we need runtime envs to ensure the example runs in CI, but we can hide that in the docs.

**Comment by Yard1** [comment]
Release test is passing https://buildkite.com/ray-project/release-tests-pr/builds/30322#0186c491-f47b-4b0a-83d5-0c3aad20e7e4

**Review by richardliaw — COMMENTED** [review]
Antoni, can you add a section that demonstrates that training is faster with more nodes/gpus?

File: doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb [hunk 2]
```diff
@@ -414,10 +414,10 @@
     "🤗 Transformers and Ray AIR's integration (`HuggingFaceTrainer`) allow you to easily configure and use DDP and DeepSpeed. All you need to do is specify the DeepSpeed configuration in the [`TrainingArguments`](https://huggingface.co/docs/transformers/en/main_classes/trainer#transformers.TrainingArguments) object.\n",
     "\n",
     "```{tip}\n",
-    "There are many DeepSpeed settings, allowing you to trade-off speed for memory usage. The settings used below are tailored to the cluster setup used (16 g4dn.4xlarge nodes) and batch size of 16. Some things to keep in mind:\n",
-    "- If your GPUs support bfloat16, it is recommended to use that instead of float16 mixed precision as it gives better performance and prevents overflows. Simply replace `fp16=True` with `bf16=True` in `TrainingArguments`\n",
+    "There are many DeepSpeed settings, allowing you to trade-off speed for memory usage. The settings used below are tailored to the cluster setup used (16 g4dn.4xlarge nodes) and per device batch size of 16. Some things to keep in mind:\n",
```
Comment (angelinalg) [inline doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb:417]:
```suggestion
    "There are many DeepSpeed settings that allow you to trade-off speed for memory usage. The settings used below are tailored to the cluster setup used (16 g4dn.4xlarge nodes) and per device batch size of 16. Some things to keep in mind:\n",
```

File: doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb [hunk 3]
```diff
@@ -414,10 +414,10 @@
...
+    "- If your GPUs support bfloat16, it is recommended to use that instead of float16 mixed precision as it gives better performance and prevents overflows. Simply replace `fp16=True` with `bf16=True` in `TrainingArguments`.\n",
```
Comment (angelinalg) [inline doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb:418]:
```suggestion
    "- If your GPUs support bfloat16, use that instead of float16 mixed precision to get better performance and prevent overflows. Replace `fp16=True` with `bf16=True` in `TrainingArguments`.\n",
```

File: doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb [hunk 4]
```diff
@@ -414,10 +414,10 @@
...
     "- If you are running out of GRAM: try reducing batch size (defined in the cell below the next one), set `\"overlap_comm\": False` in DeepSpeed config.\n",
-    "- If you are running out of RAM: try adding more nodes to your cluster, use nodes with more RAM, set `\"pin_memory\": False` in DeepSpeed config, reduce batch size and remove `\"offload_param\"` from DeepSpeed config.\n",
+    "- If you are running out of RAM: try adding more nodes to your cluster, use nodes with more RAM, set `\"pin_memory\": False` in DeepSpeed config, reduce batch size, remove `\"offload_param\"` from DeepSpeed config.\n",
```
Comment (angelinalg) [inline doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb:420]:
```suggestion
    "- If you are running out of RAM: try adding more nodes to your cluster, use nodes with more RAM. Set `\"pin_memory\": False` in the DeepSpeed config, reduce the batch size, and remove `\"offload_param\"` from the DeepSpeed config.\n",
```

File: doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb [hunk 5]
```diff
@@ -426,13 +426,13 @@
...
+    "As we are using data parallelism, each worker will operate on its own shard of the data. The batch size we set in `TrainingArguments` is the **per device batch size** (or, in other words, per worker batch size). By changing the number of workers, we can change the **effective batch size** and thus the time needed for training to complete. The effective batch size is then calculated as `per device batch size * number of workers * number of gradient accumulation steps`. As we add more workers, our effective batch size will rise and thus we will need less time to complete a full epoch. While the speedup is not exactly linear due to extra communication overheads, in many cases it can be close to linear.\n",
```
Comment (angelinalg) [inline doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb:429]:
```suggestion
    "As we are using data parallelism, each worker operates on its own shard of the data. The batch size set in `TrainingArguments` is the **per device batch size** (per worker batch size). By changing the number of workers, we can change the **effective batch size** and thus the time needed for training to complete. The effective batch size is then calculated as `per device batch size * number of workers * number of gradient accumulation steps`. As we add more workers, the effective batch size rises and thus we need less time to complete a full epoch. While the speedup is not exactly linear due to extra communication overheads, in many cases it can be close to linear.\n",
```

File: doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb [hunk 6]
```diff
@@ -426,13 +426,13 @@
...
     "\n",
     "The preprocessed dataset has 1348 examples. We have set per device batch size to 16.\n",
     "\n",
-    "* With 16 g4dn.4xlarge nodes, the effective batch size was 256, which equaled to 85 steps per epoch. One epoch took ~2440 seconds (including initialization time).\n",
+    "* With 16 g4dn.4xlarge nodes, the effective batch size was 256, which equaled to 85 steps per epoch. One epoch took **~2440 seconds** (including initialization time).\n",
```
Comment (angelinalg) [inline doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb:433]:
```suggestion
    "* With 16 g4dn.4xlarge nodes, the effective batch size was 256, which equals to 85 steps per epoch. One epoch took **~2440 seconds** (including initialization time).\n",
```

File: doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb [hunk 7]
```diff
@@ -426,13 +426,13 @@
...
     "\n",
-    "* With 32 g4dn.4xlarge nodes, the effective batch size was 512, which equaled to 43 steps per epoch. One epoch took ~1280 seconds (including initialization time)."
+    "* With 32 g4dn.4xlarge nodes, the effective batch size was 512, which equaled to 43 steps per epoch. One epoch took **~1280 seconds** (including initialization time)."
```
Comment (angelinalg) [inline doc/source/ray-air/examples/gptj_deepspeed_fine_tuning.ipynb:435]:
```suggestion
    "* With 32 g4dn.4xlarge nodes, the effective batch size was 512, which equals to 43 steps per epoch. One epoch took **~1280 seconds** (including initialization time)."
```

**Review by richardliaw — APPROVED** [review]

**Review by krfricke — APPROVED** [review]
Looks good!

File: doc/source/ray-air/examples/gptj_batch_prediction.ipynb [hunk 8]
```diff
@@ -5,11 +5,11 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# GPT-J6-B Batch Prediction with Ray AIR\n",
+    "# GPT-J-6B Batch Prediction with Ray AIR\n",
     "\n",
-    "In this example, we will showcase how to use the Ray AIR for **GPT-J batch inference**. GPT-J is a GPT-2-like causal language model trained on the Pile dataset. This particular model has 6 billion parameters.\n",
+    "In this example, we will showcase how to use the Ray AIR for **GPT-J batch inference**. GPT-J is a GPT-2-like causal language model trained on the Pile dataset. This particular model has 6 billion parameters. For more information on GPT-J, click [here](https://huggingface.co/docs/transformers/model_doc/gptj).\n",
```
Comment (angelinalg) [inline doc/source/ray-air/examples/gptj_batch_prediction.ipynb:10]:
```suggestion
    "This example showcases how to use the Ray AIR for **GPT-J batch inference**. GPT-J is a GPT-2-like causal language model trained on the Pile dataset. This model has 6 billion parameters. For more information on GPT-J, click [here](https://huggingface.co/docs/transformers/model_doc/gptj).\n",
```

File: doc/source/ray-air/examples/gptj_batch_prediction.ipynb [hunk 9]
```diff
@@ -5,11 +5,11 @@
...
     "\n",
-    "We will use Ray Data to carry out this task and a pretrained model from Hugging Face hub. Note that you can easily adapt this example to use other similar models.\n",
+    "We will use Ray Data and a pretrained model from Hugging Face hub. Note that you can easily adapt this example to use other similar models.\n",
```
Comment (angelinalg) [inline doc/source/ray-air/examples/gptj_batch_prediction.ipynb:12]:
```suggestion
    "We use Ray Data and a pretrained model from Hugging Face hub. Note that you can easily adapt this example to use other similar models.\n",
```

File: doc/source/ray-air/examples/index.rst [hunk 10]
```diff
@@ -25,6 +25,7 @@ Text/NLP
 --------
 
 - :doc:`/ray-air/examples/huggingface_text_classification`: How to use Ray AIR to run Hugging Face Transformers fine-tuning on a text classification task.
+- :doc:`/ray-air/examples/gptj_deepspeed_fine_tuning`: How to use Ray AIR to run Hugging Face Transformers with DeepSpeed for fine-tuning of a large model.
```
Comment (angelinalg) [inline doc/source/ray-air/examples/index.rst:28]:
```suggestion
- :doc:`/ray-air/examples/gptj_deepspeed_fine_tuning`: How to use Ray AIR to run Hugging Face Transformers with DeepSpeed for fine-tuning a large model.
```
